src/scalarstats.jl

```diff
@@ -260,37 +260,53 @@ variation(x) = ((m, s) = mean_and_std(x); s/m)
 realXcY(x::Real, y::Real) = x*y
 realXcY(x::Complex, y::Complex) = real(x)*real(y) + imag(x)*imag(y)
 
+
+"""
+    var_sample_mean(x[, weights::AbstractWeights]; mean=nothing)
+
+Return the estimated variance of the mean for a collection `x`. When using no weights, this is
+equal to the (sample) standard deviation divided by the sample size. If weights are used, 
+the variance of the sample mean is calculated as follows:
+
+* `AnalyticWeights`: Not implemented.
+* `FrequencyWeights`: ``\\frac{\\sum_{i=1}^n w_i (x_i - \\bar{x_i})^2}{(\\sum w_i) (\\sum w_i - 1)}``
+* `ProbabilityWeights`: ``\\frac{n}{n-1} \\frac{\\sum_{i=1}^n w_i^2 (x_i - \\bar{x_i})^2}{\\left( \\sum w_i \\right)^2}``
+
+The standard error is then the square root of the above quantities.
+"""
+
+
 """
-    sem(x)
+    sem(x[, weights::AbstractWeights]; mean=nothing)
 
-Return the standard error of the mean of collection `x`,
-i.e. `sqrt(var(x, corrected=true) / length(x))`.
+Return the standard error of the mean for a collection `x`. When using no weights, this is
+equal to the (sample) standard deviation divided by the sample size. If weights are used, 
+the variance of the sample mean is calculated as follows:
+
+* `AnalyticWeights`: Not implemented.
+* `FrequencyWeights`: ``\\frac{\\sum_{i=1}^n w_i (x_i - \\bar{x_i})^2}{(\\sum w_i) (\\sum w_i - 1)}``
+* `ProbabilityWeights`: ``\\frac{n}{n-1} \\frac{\\sum_{i=1}^n w_i^2 (x_i - \\bar{x_i})^2}{\\left( \\sum w_i \\right)^2}``
+
+The standard error is then the square root of the above quantities.
 """
-function sem(x)
-    y = iterate(x)
-    if y === nothing
-        T = eltype(x)
-        # Return the NaN of the type that we would get, had this collection
-        # contained any elements (this is consistent with std)
-        return oftype(sqrt((abs2(zero(T)) + abs2(zero(T)))/2), NaN)
+sem(x; mean=nothing) = sqrt(var(x; mean=mean, corrected=true) / length(x))
+
+function sem(x, weights::UnitWeights; mean=nothing)
+    length(x) != length(weights) && throw(DimensionMismatch("inconsistent array dimension"))
+    return sem(x; mean=mean)
+end
     end
-    count = 1
-    value, state = y
-    y = iterate(x, state)
-    # Use Welford algorithm as seen in (among other places)
-    # Knuth's TAOCP, Vol 2, page 232, 3rd edition.
-    M = value / 1
-    S = real(zero(M))
-    while y !== nothing
-        value, state = y
-        y = iterate(x, state)
-        count += 1
-        new_M = M + (value - M) / count
-        S = S + realXcY(value - M, value - new_M)
-        M = new_M
+end
+
+# Weighted methods for the above
+sem(x, weights::FrequencyWeights) = sqrt(var(x, weights; corrected=true) / weights.sum)
+
+function sem(x, weights::ProbabilityWeights)
+    μ = mean(x, weights)
+    var = sum(zip(x, weights)) do (x, w)
```

Comment:
Is this supposed to be the weighted variance for `ProbabiltyWeights` `weights`? I wonder because you use the name `var` but it does not correspond to `var(x, weights; mean=mean)`.

Reply:
> Is this supposed to be the weighted variance for `ProbabiltyWeights` `weights`? I wonder because you use the name `var` but it does not correspond to `var(x, weights; mean=mean)`.

It's not; it's the variance of the sample mean before normalizing.

Reply:
In terms of implementation, something like this would be a bit faster and more accurate I think as it will use pairwise summation:
```julia
sum(Base.Broadcast.broadcasted(weights, x) do w, xi
    (w * (xi - mean))^2
end)
(We really need a better syntax for this...)

Reply:
Then we should restrict `x` to `AbstractArray` or `AbstractArray{<:Real}`  and use `vec(x)` I assume.

Reply:
```suggestion
sum(Base.Broadcast.broadcasted(weights, vec(x)) do w, xi
    (w * (xi - mean))^2
end)
```

Reply:
Yes in general we currently require arrays when passing weights so that the correspondence between data and weights is clear.

Reply:
> In terms of implementation, something like this would be a bit faster and more accurate I think as it will use pairwise summation:
> 
> ```julia
> sum(Base.Broadcast.broadcasted(weights, x) do w, xi
>     (w * (xi - mean))^2
> end)
> (We really need a better syntax for this...)
> ```

Not sure I understand, what's the difference?

Reply:
The difference is that `Broadcasted` objects use the same summation algorithm as arrays, i.e. pairwise summation. This is is more precise, and sometimes even faster thanks to `@inbounds`. While this function indeed doesn't have a docstring, it is documented [in the manual](https://docs.julialang.org/en/v1/manual/interfaces/#man-interfaces-broadcasting).